### PR TITLE
Implement stop request in federated execution of TypeScript

### PR DESCRIPTION
**Pull request by ByeongGil-Jun**
Relevant issue: https://github.com/lf-lang/reactor-ts/issues/101

Relevant LF PR: https://github.com/lf-lang/lingua-franca/pull/1302

Add a logic for sending and propagating MSG_TYPE_STOP_REQUEST and MSG_TYPE_STOP_REQUEST_REPLY and receive MSG_TYPE_STOP_REQUEST_GRANTED
Discussion is needed to determine whether freeze federates or not.
Some function in reactor.ts is added and modified like getStartTag() and getStartTime() etc.

### CODE REVIEW

File: src/core/federation.ts
Comment (lhstrh) [inline src/core/federation.ts:154]: This is not the right place to put this. How about we create a wiki page and link to it?

**Review by lhstrh — CHANGES_REQUESTED** [review]
Looks good to me! My only request is that we start putting developer docs about federated execution on the wiki.

File: src/core/federation.ts
Comment (hokeun) [reply src/core/federation.ts]: Done win https://github.com/lf-lang/reactor-ts/pull/111/commits/b399e4fa9de86f341d1313af173450b79b225ce1!

@lhstrh, How do you think about we add @ByeongGil-Jun as a member of LF repo so that he can edit the wiki next time?

**Comment by hokeun** [comment]
@ByeongGil-Jun, I granted you write access to the lingua-franca repo (https://github.com/lf-lang/lingua-franca/). From now on, you should be able to create branches and edit wiki pages here: https://github.com/lf-lang/lingua-franca/wiki

**Review by hokeun — APPROVED** [review]
I'm approving this to unblock @ByeongGil-Jun. @lhstrh, please feel free to comment on this PR if you come up with any suggestions later!

**Pull request merged**
